src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,301 @@
+use super::super::SinsemillaInstructions;
+use super::{get_s_by_idx, CellValue, EccPoint, SinsemillaChip, Var};
+use crate::circuit::gadget::utilities::copy;
+use crate::primitives::sinsemilla::lebs2ip_k;
+use halo2::{
+    arithmetic::{CurveAffine, FieldExt},
+    circuit::{Chip, Region},
+    plonk::Error,
+};
+
+use ff::{Field, PrimeFieldBits};
+use group::Curve;
+use pasta_curves::pallas;
+
+#[cfg(test)]
+use group::prime::PrimeCurveAffine;
+
+impl<const K: usize, const MAX_WORDS: usize> SinsemillaChip<K, MAX_WORDS> {
+    #[allow(non_snake_case)]
+    pub(super) fn hash_message(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        Q: pallas::Affine,
+        message: &<Self as SinsemillaInstructions<pallas::Affine, K, MAX_WORDS>>::Message,
+    ) -> Result<EccPoint, Error> {
+        let config = self.config().clone();
+
+        // Get the `x`- and `y`-coordinates of the starting `Q` base.
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Assign the `x`-coordinate of the starting `Q` base.
+        let x_q_cell = region.assign_advice(|| "x_q", config.x_a, 0, || Ok(x_q))?;
+
+        // Initialize the accumulator to `Q`.
+        let mut x_a = CellValue::new(x_q_cell, Some(x_q));
+        let mut y_a = Some(y_q);
+
+        // Hash each piece in the message.
+        let mut offset = 0;
+        for piece in message.0.iter() {
+            // The value of the hash after this piece is processed.
+            let a = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words();
+
+            // Update the accumulator to the latest hash value.
+            x_a = a.0;
+            y_a = a.1;
+        }
+
+        // Assign the final `y_a`.
+        let y_a_cell = region.assign_advice(
+            || "y_a",
+            config.word,
+            offset + 1,
+            || y_a.ok_or(Error::SynthesisError),
+        )?;
```

Comment:
`y_a` is assigned from the return value of `hash_piece` in the last step.

Reply:
The x-coordinate is not sufficient to define the point, because sign is significant. It should be constrained by the last constraint in https://hackmd.io/iOw7-HpFQY6dPF1aFY8pAw#Constraint-program :

λ_{2,n} ⋅ (x_{A,n} − x_{A,n+1}) = (λ_{1,n} + λ_{2,n}) ⋅ (x_{A,n} − (λ_{1,n}^2 − x_{A,n} − x_{P,n})) + y_{A,n+1}

Reply:
This constraint is called "Sinsemilla gate 2" in this PR.

Reply:
Resolved by the new constraints.